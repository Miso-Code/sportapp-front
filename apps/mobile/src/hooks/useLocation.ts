import React, { useState, useEffect } from 'react'
import * as Location from 'expo-location'

const startLocationUpdates = async (
	setIsLocationAvailable: React.Dispatch<React.SetStateAction<boolean>>,
	setLocationUpdates: React.Dispatch<
		React.SetStateAction<Location.LocationObject[]>
	>
) => {
	try {
		let { status } = await Location.requestForegroundPermissionsAsync()
		if (status !== 'granted') {
			setIsLocationAvailable(false)
			return
		}

		setIsLocationAvailable(true)

		await Location.startLocationUpdatesAsync('locationUpdates', {
			accuracy: Location.Accuracy.High,
			timeInterval: 1000,
			distanceInterval: 1
		})

		Location.watchPositionAsync(
			{
				accuracy: Location.Accuracy.High,
				timeInterval: 1000
			},
			(location) => {
				setLocationUpdates((prevLocations) => [
					...prevLocations,
					location
				])
			}
		)
	} catch (error) {
		console.error('Error starting location updates', error) // eslint-disable-line no-console
	}
}

const stopLocationUpdates = async () => {
	await Location.stopLocationUpdatesAsync('locationUpdates')
}

export const useLocation = () => {
	const [locationUpdates, setLocationUpdates] = useState<
		Location.LocationObject[]
	>([])
	const [isLocationAvailable, setIsLocationAvailable] = useState(false)

	useEffect(() => {
		let isMounted = true

		if (isMounted) {
			startLocationUpdates(setIsLocationAvailable, setLocationUpdates)
		}

		return () => {
			isMounted = false
			stopLocationUpdates()
		}
	}, [])

	return { locationUpdates, isLocationAvailable }
}
