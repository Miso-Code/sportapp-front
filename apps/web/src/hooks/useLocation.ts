import { useEffect, useState } from 'react'

export const useLocation = () => {
	const [location, setLocation] = useState({
		latitude: 0,
		longitude: 0
	})
	const [error, setError] = useState('')
	const [loading, setLoading] = useState(true)

	const handleSuccess = (position: GeolocationPosition) => {
		setLocation({
			latitude: position.coords.latitude,
			longitude: position.coords.longitude
		})
	}

	const handleError = (error: GeolocationPositionError) => {
		if (error.code === 1) {
			setError('User denied the request for Geolocation')
		} else if (error.code === 2) {
			setError('Location information is unavailable')
		} else if (error.code === 3) {
			setError('The request to get user location timed out')
		} else {
			setError('An unknown error occurred')
		}
	}

	useEffect(() => {
		if ('geolocation' in navigator) {
			navigator.geolocation.getCurrentPosition(handleSuccess, handleError)
		} else {
			setError('Geolocation is not available')
		}
		setLoading(false)
	}, [])

	return { location, error, loading }
}
