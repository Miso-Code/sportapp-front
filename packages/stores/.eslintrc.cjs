const eslintDefaultConfig = require('@sportapp/eslint-config/library')

module.exports = {
	...eslintDefaultConfig
}
