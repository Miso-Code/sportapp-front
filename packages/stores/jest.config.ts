export default {
	preset: 'ts-jest',
	setupFilesAfterEnv: ['<rootDir>/config/test/jest.setup.ts'],
	testEnvironment: 'jsdom',
	transform: {
		'^.+\\.tsx?$': ['ts-jest']
	},
	moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
	collectCoverage: true,
	collectCoverageFrom: [
		'<rootDir>/src/**/*.{ts,tsx}',
		'!<rootDir>/src/**/interfaces/**/*.{ts,tsx}',
		'!<rootDir>/src/**/__mocks__/**/*.{ts,tsx}'
	],
	moduleNameMapper: {
		'\\.(gif|ttf|eot|svg|png|jpg)$':
			'<rootDir>/config/test/__mocks__/fileMock.ts',
		'^.+\\.(css|less|scss)$': 'babel-jest',
		'^@/(.*)$': ['<rootDir>/src/$1']
	},
	coverageThreshold: {
		global: {
			branches: 70,
			functions: 70,
			lines: 70,
			statements: 70
		}
	},
	reporters: ['default']
}
