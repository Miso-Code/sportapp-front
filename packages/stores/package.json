{
	"name": "@sportapp/stores",
	"version": "1.0.0",
	"license": "MIT",
	"main": "src/index.ts",
	"type": "module",
	"scripts": {
		"format": "prettier --write \"**/*.{ts,tsx,md,js}\"",
		"lint": "yarn lint:js",
		"lint:js": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
		"pre-commit": "lint-staged",
		"test": "jest --ci --reporters=default --config jest.config.ts --no-cache --coverage --forceExit --detectOpenHandles --logHeapUsage --silent=false --verbose"
	},
	"dependencies": {
		"@react-native-async-storage/async-storage": "^1.23.1",
		"@sportapp/sportapp-repository": "*",
		"@tanstack/react-query": "^5.28.9",
		"@testing-library/jest-dom": "^6.4.2",
		"@testing-library/react": "^14.2.2",
		"@types/jest": "^29.5.12",
		"babel-jest": "^29.7.0",
		"jest": "^29.7.0",
		"prettier": "^3.2.5",
		"react": "^18.2.0",
		"ts-jest": "^29.1.2",
		"ts-jest-mock-import-meta": "^1.2.0",
		"ts-node": "^10.9.2",
		"zustand": "^4.5.2"
	},
	"devDependencies": {
		"@sportapp/eslint-config": "*",
		"@sportapp/prettier-config": "*",
		"@sportapp/typescript-config": "*",
		"eslint": "^8.19.0",
		"lint-staged": "^15.2.2",
		"simple-zustand-devtools": "^1.1.0"
	}
}
