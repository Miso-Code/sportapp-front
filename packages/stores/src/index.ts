export { useAuthStore, initialAuthState } from './auth'
export { useSportSessionStore, initialSportSessionState } from './sportSession'
export { useSportStore, initialSportState } from './sport'
export { useUserStore, initialUserState } from './user'
export {
	usePartnerAuthStore,
	initialAuthState as initialAuthStatePartner
} from './partner/auth'
export {
	useBusinessPartnerStore,
	initialBusinessPartnerState
} from './businessPartner'
export { useAlertStore, initialAlertState } from './alert'
export {
	initialProductPartnerState,
	usePartnerProductStore
} from './partner/products'

export { initialTrainingPlanState, useTrainingPlanStore } from './trainingPlan'

export { initialPaymentPlanState, usePaymentPlanStore } from './paymentPlan'

export { initialSportsmanState, useSportsmanStore } from './sportsman'

export { initialSportEventState, useSportEventStore } from './sportEvents'
