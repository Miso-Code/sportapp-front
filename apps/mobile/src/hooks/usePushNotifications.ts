// Remove this line after implementing the code
import { useCallback, useState } from 'react'
import messaging from '@react-native-firebase/messaging'
import { PermissionsAndroid, Platform } from 'react-native'
import { useAlertStore } from '@sportapp/stores'

type ForegroundNotificationCallback = (
	title: string,
	message: string,
	type: 'info' | 'warning'
) => void

export const usePushNotification = () => {
	const [isPermissionsGranted, setIsPermissionsGranted] = useState(false)
	const { addHiddenAlertToHistory } = useAlertStore()

	const subscribe = useCallback(
		async (callback?: ForegroundNotificationCallback) => {
			await Promise.all([
				listenToForegroundNotifications(callback),
				listenToBackgroundNotifications(),
				onNotificationOpenedAppFromBackground(),
				onNotificationOpenedAppFromQuit()
			])
		},
		// eslint-disable-next-line react-hooks/exhaustive-deps
		[]
	)

	const getToken = useCallback(async () => {
		if (!isPermissionsGranted) await requestUserPermission()
		return await getFCMToken()
	}, [isPermissionsGranted])

	const requestUserPermission = async () => {
		if (Platform.OS === 'ios') {
			//Request iOS permission
			const authStatus = await messaging().requestPermission()
			const enabled =
				authStatus === messaging.AuthorizationStatus.AUTHORIZED ||
				authStatus === messaging.AuthorizationStatus.PROVISIONAL

			setIsPermissionsGranted(enabled)
		} else if (Platform.OS === 'android') {
			//Request Android permission (For API level 33+, for 32 or below is not required)
			const status = await PermissionsAndroid.request(
				PermissionsAndroid.PERMISSIONS.POST_NOTIFICATIONS
			)
			const enabled = status === PermissionsAndroid.RESULTS.GRANTED

			setIsPermissionsGranted(enabled)
		}
	}

	const getFCMToken = async () => {
		try {
			return await messaging().getToken()
		} catch (error) {
			// eslint-disable-next-line no-console
			console.error('There was a problem getting FCM token', error)
		}
	}

	const listenToForegroundNotifications = async (
		callback: ForegroundNotificationCallback = () => {}
	) => {
		const unsubscribe = messaging().onMessage(async (remoteMessage) => {
			callback(
				remoteMessage.notification.title,
				remoteMessage.notification.body,
				remoteMessage.data.priority as 'info' | 'warning'
			)
		})
		return unsubscribe
	}

	const listenToBackgroundNotifications = async () => {
		const unsubscribe = messaging().setBackgroundMessageHandler(
			async (remoteMessage) => {
				addHiddenAlertToHistory({
					type: remoteMessage.data.priority as 'info' | 'warning',
					message: `${remoteMessage.notification.title}: ${remoteMessage.notification.body}`
				})
			}
		)
		return unsubscribe
	}

	const onNotificationOpenedAppFromBackground = async () => {
		const unsubscribe = messaging().onNotificationOpenedApp(
			async (remoteMessage) => {
				addHiddenAlertToHistory({
					type: remoteMessage.data.priority as 'info' | 'warning',
					message: `${remoteMessage.notification.title}: ${remoteMessage.notification.body}`
				})
			}
		)
		return unsubscribe
	}

	const onNotificationOpenedAppFromQuit = async () => {
		const remoteMessage = await messaging().getInitialNotification()

		if (remoteMessage) {
			addHiddenAlertToHistory({
				type: remoteMessage.data.priority as 'info' | 'warning',
				message: `${remoteMessage.notification.title}: ${remoteMessage.notification.body}`
			})
		}
	}

	return {
		isPermissionsGranted,
		subscribe,
		getToken
	}
}
