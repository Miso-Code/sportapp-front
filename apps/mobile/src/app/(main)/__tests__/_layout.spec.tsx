import React from 'react'
import renderer, { ReactTestRenderer, act } from 'react-test-renderer'

import { useAlertStore } from '@sportapp/stores'

import AppLayout from '../_layout'
import { usePushNotification } from '@/hooks/usePushNotifications'

jest.mock('expo-router/stack')
jest.mock('react-native-paper', () => ({
	...jest.requireActual('react-native-paper'),
	PaperProvider: jest.fn(({ children }) => children),
	Portal: jest.fn(({ children }) => children)
}))

jest.mock('@sportapp/stores', () => ({
	useAlertStore: jest.fn().mockReturnValue({
		alert: undefined,
		setAlert: jest.fn()
	})
}))

jest.mock('react-native-safe-area-context', () => ({
	useSafeAreaInsets: () => ({
		bottom: 0,
		top: 0,
		right: 0,
		left: 0
	}),
	SafeAreaProvider: ({ children }: { children: React.ReactNode }) => children
}))

jest.mock('@/hooks/usePushNotifications', () => ({
	usePushNotification: jest.fn().mockReturnValue({ subscribe: jest.fn() })
}))

describe('AppLayout', () => {
	let component: ReactTestRenderer

	beforeEach(() => {
		jest.clearAllMocks()
		component = renderer.create(<AppLayout />)
	})

	afterEach(() => {
		component.unmount()
	})

	it('should render correctly', () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should subscribe to push notifications on mount', async () => {
		await act(async () => {
			await Promise.resolve()
		})
		expect(usePushNotification().subscribe).toHaveBeenCalled()
	})

	it('should render an snackbar on alert', () => {
		;(useAlertStore as unknown as jest.Mock).mockReturnValue({
			alert: {
				message: 'test',
				type: 'info'
			},
			setAlert: jest.fn()
		})
		component.update(<AppLayout />)
		const [snack] = component.root.findAllByProps({ testID: 'alert' })
		expect(snack).toBeDefined()
	})

	it('should unset alert on snackbar dismiss', () => {
		;(useAlertStore as unknown as jest.Mock).mockReturnValue({
			alert: {
				message: 'test',
				type: 'info'
			},
			setAlert: jest.fn()
		})
		component.update(<AppLayout />)
		const [snack] = component.root.findAllByProps({ testID: 'alert' })
		snack.props.onDismiss()
		expect(useAlertStore().setAlert).toHaveBeenCalledWith(undefined)
	})

	it('should unset alert on snackbar action', () => {
		;(useAlertStore as unknown as jest.Mock).mockReturnValue({
			alert: {
				message: 'test',
				type: 'info',
				action: {
					label: 'test',
					onPress: jest.fn()
				}
			},
			setAlert: jest.fn()
		})
		component.update(<AppLayout />)
		const [snack] = component.root.findAllByProps({ testID: 'alert' })
		snack.props.onIconPress()
		expect(useAlertStore().setAlert).toHaveBeenCalledWith(undefined)
	})

	it('should not render an snackbar on no alert', () => {
		;(useAlertStore as unknown as jest.Mock).mockReturnValue({
			alert: undefined,
			setAlert: jest.fn()
		})
		component.update(<AppLayout />)
		const snack = component.root.findAllByProps({ testID: 'alert' })
		expect(snack).toHaveLength(0)
	})

	it('should show an info message on push notification', async () => {
		await act(async () => {
			await Promise.resolve()
		})

		const showPushNotification = (
			usePushNotification().subscribe as jest.Mock
		).mock.calls[0][0]

		act(() => showPushNotification('test', 'test', 'info'))

		expect(useAlertStore().setAlert).toHaveBeenCalledWith({
			message: 'test',
			type: 'info',
			position: 'top'
		})
	})

	it('should show a warning message on push notification', async () => {
		await act(async () => {
			await Promise.resolve()
		})

		const showPushNotification = (
			usePushNotification().subscribe as jest.Mock
		).mock.calls[0][0]
		showPushNotification('test', 'test', 'warning')
		expect(useAlertStore().setAlert).toHaveBeenCalledWith({
			message: 'test',
			type: 'warning',
			position: 'top'
		})
	})
})
