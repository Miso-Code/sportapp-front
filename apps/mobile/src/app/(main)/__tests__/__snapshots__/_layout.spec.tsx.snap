// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLayout should render correctly 1`] = `
<View>
  <View>
    <Text>
      Mock Screen
    </Text>
  </View>
</View>
`;
